modules/create_timepoint/php/timepoint.class.inc

```diff
@@ -0,0 +1,344 @@
+            unset($values['psc']);
+        }
+
+        return new \LORIS\Http\Response\JsonResponse(
+            $values
+        );
+    }
+
+    /**
+     * Processes the values & saves to database and return a json response.
+     *
+     * @param ServerRequestInterface $request The incoming PSR7 request.
+     *
+     * @return ResponseInterface The outgoing PSR7 response
+     */
+    private function _handlePOST(ServerRequestInterface $request) : ResponseInterface
+    {
+        // Parse POST request body.
+        $values = json_decode((string) $request->getBody(), true);
+
+        if (!$this->_verifyCreatePermissions($request, $values['identifier'])) {
+            return new \LORIS\Http\Response\JSON\Forbidden(
+                'Forbidden'
+            );
+        }
+
+        // Validate POST data.
+        $values = $this->_validate($values);
+
+        // convert site entered to a \Site object
+        $user = \User::singleton();
+
+        $user_list_of_sites = $user->getCenterIDs();
+        $num_sites          = count($user_list_of_sites);
+        $site = null;
+        if ($num_sites == 1) {
+            $site = \Site::singleton($user_list_of_sites[0]);
+        } else if ($num_sites > 1) {
+            $site = \Site::singleton(intval($values['psc']));
+        }
+
+        // Project
+        $project = null;
+        $user_list_of_projects = $user->getProjectIDs();
+        if (count($user_list_of_projects) == 1) {
+            // if there is only one project, autoselect first project from array of 1
+            $project = \Project::getProjectFromID(array_pop($user_list_of_projects));
+        } else if (count($user_list_of_projects) > 1) {
+            $project_id = intval($values['psc']);
+            $project    = \Project::getProjectFromID($project_id);
+        }
+
+        if (isset($values['conflict'])) {
+            return new \LORIS\Http\Response\JSON\Conflict(
+                json_encode($values['conflict'])
+            );
+        }
+        \TimePoint::createNew(
+            \Candidate::singleton(new CandID($values['candID'])),
+            intval($values['subproject']),
+            $values['visit'] ?? null,
+            $site,
+            $project,
+            intval($values['languages']) ?? null
+        );
+        return new \LORIS\Http\Response\JSON\Created(['success']);
+    }
+
+    /**
+     * Permission verification.
+     *
+     * @param ServerRequestInterface $request    The incoming PSR7 request.
+     * @param string                 $identifier The identifier.
+     *
+     * @return bool
+     */
+    private function _verifyCreatePermissions(
+        ServerRequestInterface $request,
+        string $identifier
+    ) : bool {
+        $user      = $request->getAttribute('user');
+        $candidate = \Candidate::singleton(new CandID($identifier));
+        return (
+            $user->hasPermission('data_entry') &&
+            $candidate->getData('RegistrationCenterID')
+        );
```

Comment:
should this maybe be the same check as the access function in create_timepoint.class.inc ? i think we said that as long as the user has access to the timepoint_list page and has data_entry permissions, they should be able to createa  time point. is this right @driusan @ridz1208?

Reply:
@zaliqarosli I pasted the wrong line on 229 from your fix.
It used to be:
```
return (
            $user->hasPermission('data_entry') &&
            (in_array(
                $candidate->getData('RegistrationCenterID'),
                $user->getData('CenterIDs')
            )
            )
        );
```
should be similar to what you proposed. I'll fix in latest commit.

Reply:
@zaliqarosli ready

Reply:
hey @maltheism sorry for suggesting this change! turns out `$this->candidate->isAccessibleBy($user)` isn't exactly the same as what was there before. before, only if you were affiliated with the candidate's registration centre were you able to create the timepoint. using the isAccessibleBy method allows you to create a timepoint if you aren't affiliated with the candidate's registration centre. could you please change it back? sorry about that

Reply:
@zaliqarosli done reverted changes.